dagmc_stats/DagmcQuery.py

```diff
@@ -236,3 +303,268 @@ def calc_area_triangle(self, tris=[]):
             tri_area.append(row_data)
         self.__update_tri_data(tri_area)
+
+    def calc_coarseness(self):
+        """
+        Calculate the surface coarseness data
+
+        inputs
+        ------
+            none
+
+        outputs
+        -------
+            none
+        """
+        if 'coarseness' in self._surf_data:
+            warnings.warn('Coarseness already exists. ' +
+                          'Calc_coarseness() will not be called.')
+            return
+
+        coarseness = []
+        surfs = self.get_surfs()
+        self.calc_area_triangle()
+        total_area = 0.
+        weighted_coarseness = 0.  # surf coarseness * surf area
+        for surf in surfs:
+            tris = self.dagmc_file._my_moab_core.get_entities_by_type(
+                surf, types.MBTRI)
+            area = self._tri_data.loc[self._tri_data['tri_eh'].isin(
+                list(tris)), 'area'].sum()
+            cval = len(tris) / area
+            row_data = {'surf_eh': surf, 'coarseness': cval}
+            coarseness.append(row_data)
+            total_area += area
+            weighted_coarseness += cval * area
+        self.__update_surf_data(coarseness)
+
+        # calculate weighted average
+        # = sum(coarseness_i * area_i) / sum(area_i), i = surface
+        average_coarseness = weighted_coarseness / total_area
+        self._global_averages['coarseness_ave'] = average_coarseness
+
+    def __get_tri_vert_data(self):
+        """Build a numpy strcutured array to store triangle and vertex related
+        data in the form of triangle entity handle | vertex entity handle
+        | angle connected to the vertex | side length of side opposite to
+        vertex in the triangle
+
+        inputs
+        ------
+        none
+
+        outputs
+        -------
+        tri_vert_data : a numpy structured array that stores the triangle
+            and vertex related data
+        all_verts : (list) all the vertices that are connected to
+            triangle in the geometry
+        """
+        all_verts = set()
+        tri_vert_struct = np.dtype({
+            'names': ['tri', 'vert', 'angle', 'side_length'],
+            'formats': [np.uint64, np.uint64, np.float64, np.float64]})
+        tris = self.get_tris()
+        tri_vert_data = np.zeros(len(tris) * 3, dtype=tri_vert_struct)
+        tri_vert_index = 0
+
+        for tri in tris:
+            side_lengths = self.get_tri_side_length(tri)  # {vert: side_length}
+            side_length_sum_sq_half = sum(map(lambda i: i**2,
+                                          side_lengths.values())) / 2.
+            side_length_prod = np.prod(list(side_lengths.values()))
+            verts = list(self.dagmc_file._my_moab_core.get_adjacencies(
+                tri, 0, op_type=1))
+            for vert_i in verts:
+                all_verts.add(vert_i)
+                side_i = side_lengths[vert_i]
+                d_i = np.arccos((side_length_sum_sq_half -
```

Comment:
I don't think we want to do that just because `side_lengths` is a dictionary where the key is the vert id. If we make that a numpy array instead we lose that information which we need below. I think that is the reason we have it in this for-loop right now.